src/tpm2_wrap.c

```diff
@@ -7123,4 +7192,321 @@ int wolfTPM2_SetIdentityAuth(WOLFTPM2_DEV* dev, WOLFTPM2_HANDLE* handle,
 
 
 
+
+/******************************************************************************/
+/* --- BEGIN Firmware Upgrade Support -- */
+/******************************************************************************/
+
+#ifdef WOLFTPM_FIRMWARE_UPGRADE
+#if defined(WOLFTPM_SLB9672) || defined(WOLFTPM_SLB9673)
+
+/* Maximum size of firmware chunks */
+#define IFX_FW_MAX_CHUNK_SZ 1024
+
+/* Setup the policy to enable firmware upgrade start */
+static int tpm2_ifx_firmware_enable_policy(WOLFTPM2_DEV* dev)
+{
+    int rc;
+    PolicyCommandCode_In policyCC;
+    SetPrimaryPolicy_In policy;
+    WOLFTPM2_SESSION tpmSession;
+
+    XMEMSET(&tpmSession, 0, sizeof(tpmSession));
+    XMEMSET(&policyCC, 0, sizeof(policyCC));
+    XMEMSET(&policy, 0, sizeof(policy));
+
+    rc = wolfTPM2_StartSession(dev, &tpmSession, NULL, NULL,
+        TPM_SE_POLICY, TPM_ALG_NULL);
+    if (rc == TPM_RC_SUCCESS) {
+        policyCC.policySession = tpmSession.handle.hndl;
+        policyCC.code = TPM_CC_FieldUpgradeStartVendor;
+        rc = TPM2_PolicyCommandCode(&policyCC);
+        if (rc == TPM_RC_SUCCESS) {
+            word32 policySz = (word32)sizeof(policy.authPolicy.buffer);
+            rc = wolfTPM2_GetPolicyDigest(dev, tpmSession.handle.hndl,
+                policy.authPolicy.buffer, &policySz);
+            policy.authPolicy.size = policySz;
+        }
+        wolfTPM2_UnloadHandle(dev, &tpmSession.handle);
+    }
+    if (rc == TPM_RC_SUCCESS) {
+        policy.authHandle = TPM_RH_PLATFORM;
+        policy.hashAlg = TPM_ALG_SHA256;
+        rc = TPM2_SetPrimaryPolicy(&policy);
+    }
+
+#ifdef DEBUG_WOLFTPM
+    if (rc != TPM_RC_SUCCESS) {
+        printf("Enable firmware start policy failed 0x%x: %s\n",
+            rc, TPM2_GetRCString(rc));
+    }
+#endif
+    return rc;
+}
+
+static int tpm2_ifx_firmware_start(WOLFTPM2_DEV* dev, TPM_ALG_ID hashAlg,
+    uint8_t* manifest_hash, uint32_t manifest_hash_sz)
+{
+    int rc;
+    WOLFTPM2_SESSION tpmSession;
+    PolicyCommandCode_In policyCC;
+
+    XMEMSET(&tpmSession, 0, sizeof(tpmSession));
+    XMEMSET(&policyCC, 0, sizeof(policyCC));
+
+    rc = wolfTPM2_StartSession(dev, &tpmSession, NULL, NULL,
+        TPM_SE_POLICY, TPM_ALG_NULL);
+    if (rc == TPM_RC_SUCCESS) {
+        policyCC.policySession = tpmSession.handle.hndl;
+        policyCC.code = TPM_CC_FieldUpgradeStartVendor;
+        rc = TPM2_PolicyCommandCode(&policyCC);
+        if (rc == TPM_RC_SUCCESS) {
+            /* build command for manifest header */
+            uint16_t val16;
+            uint8_t cmd[1 + 2 + 2 + TPM_SHA512_DIGEST_SIZE];
```

Comment:
Nitpick, but comments explaining the `cmd[]` sizing in these functions might be nice.

I'm guessing this is type (1) + data size (2) + hash alg (2).